[Quant] Add int8 linear op impl for quantization PT2E with Inductor. input is an int8 CPU tensor; weight is an int8 MdkldnnCPU tensor.

Xia-Weiwen opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #106935
* #106934
* #106782
* #106781
* __->__ #105818

**Summary**
Add a new onednn qlinear op for quantization PT2E with Inductor. input is an int8 CPU tensor and weight is an int8 MkldnnCPU tensor.

**Test plan**
python test/test_quantization.py -k test_qlinear_pt2e

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

Xia-Weiwen committed:
[Quant] Add int8 linear op impl for quantization PT2E with Inductor. input is an int8 CPU tensor; weight is an int8 MdkldnnCPU tensor.

[ghstack-poisoned]

jgong5 commented on aten/src/ATen/native/quantized/cpu/qlinear.cpp:
```diff
@@ -900,6 +901,96 @@ at::Tensor PackedLinearWeightsOnednn:: apply_tanh(
+    at::Tensor input, // int8 CPU Tensor, not QTensor
+    double input_scale,
+    int64_t input_zero_point,
+    at::Tensor mkldnn_weight, // int8 tensor from MkldnnCPU
```
can we use consistent names, i.e., `onednn_weight`?

Xia-Weiwen replied:
Thanks. It's fixed.

jgong5 commented on aten/src/ATen/native/quantized/cpu/qlinear_prepack.cpp:
```diff
@@ -383,6 +403,19 @@ class QLinearPackWeightFp16Legacy final {
+ public:
+  static at::Tensor run(
+    at::Tensor weight, // Not QTensor
+    torch::List<int64_t> input_shape) {
```
Is `input_shape` required or not? I think it is optional? Can we clarify this in the comment?

Xia-Weiwen committed:
Update on "[Quant] Add int8 linear op impl for quantization PT2E with Inductor. input is an int8 CPU tensor; weight is an int8 MdkldnnCPU tensor."




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant] Add int8 linear op impl for quantization PT2E with Inductor. input is an int8 CPU tensor; weight is an int8 MdkldnnCPU tensor."




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant] Add int8 linear op impl for quantization PT2E with Inductor. input is an int8 CPU tensor; weight is an int8 MdkldnnCPU tensor."




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant] Add int8 linear op impl for quantization PT2E with Inductor. input is an int8 CPU tensor; weight is an int8 MdkldnnCPU tensor."




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant] Add int8 linear op impl for quantization PT2E with Inductor. input is an int8 CPU tensor; weight is an int8 MdkldnnCPU tensor."




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant] Add int8 linear op impl for quantization PT2E with Inductor. input is an int8 CPU tensor; weight is an int8 MdkldnnCPU tensor."




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel commented on aten/src/ATen/native/quantized/cpu/qlinear.cpp:
```diff
@@ -900,6 +901,104 @@ at::Tensor PackedLinearWeightsOnednn:: apply_tanh(
+        output_scale == 1.0f && output_zero_point == 0, "onednn qlinear: expect scale=1 and zero point=0 for fp32 output");
+  }
+
+  auto src = at::native::itensor_from_tensor(input);
```
Do we need to ensure the input is in `contiguous` format before using its data ptr?

Xia-Weiwen replied:
Thanks. It's fixed.

leslie-fang-intel commented on aten/src/ATen/native/quantized/cpu/qlinear.cpp:
```diff
@@ -900,6 +901,104 @@ at::Tensor PackedLinearWeightsOnednn:: apply_tanh(
+    at::Tensor weight_scales,
+    at::Tensor weight_zero_points,
+    c10::optional<at::Tensor> bias, // plain tensor
+    double output_scale,
```
After inductor freezing pass, the `output_scale` from the quant node at output should be reciprocal. Do we need to do the reciprocal here to convert it back?

Xia-Weiwen replied:
Thanks. It's fixed.

leslie-fang-intel commented on aten/src/ATen/native/quantized/cpu/qlinear.cpp:
```diff
@@ -900,6 +901,104 @@ at::Tensor PackedLinearWeightsOnednn:: apply_tanh(
+      tensor::desc(onednn_bias.value().get_dims(), ideep::data_type::f32, ideep::format_tag::any) :
+      tensor::desc();
+  auto op_attr = onednn_utils::create_attr_by_post_op(post_op_name, post_op_args);
+  op_attr.set_scales_mask(DNNL_ARG_SRC, 0);
```
Similar to the oneDNN 3.1 upgrade, do we need to check the scale is not 1.0 and zp is not 0 before setting the mask? Otherwise, will it be slow as the issue we met in oneDNN 3.1 upgrade?

Xia-Weiwen replied:
Thanks. Checks are added.

leslie-fang-intel commented on aten/src/ATen/native/quantized/cpu/qlinear_prepack.cpp:
```diff
@@ -281,6 +282,25 @@ c10::intrusive_ptr<LinearPackedParamsBase> PackedLinearWeightsOnednn::prepack(
+  ideep::tensor wei = ideep::tensor({w_dims, dnnl::memory::data_type::s8}, weight.data_ptr());
+  wei.transpose_(0, 1); // oneDNN requires transposed weight
+  ideep::dims input_dims = input_shape.has_value() ? input_shape.value().vec() : ideep::dims();
+  auto w_desc = ideep::matmul_forward::expected_weights_desc(
```
Do we need to pass in the input zp and scale to query the weight format?

Xia-Weiwen replied:
Thanks. It's fixed.

leslie-fang-intel commented on aten/src/ATen/native/quantized/cpu/qlinear.cpp:
```diff
@@ -900,6 +901,104 @@ at::Tensor PackedLinearWeightsOnednn:: apply_tanh(
+  auto primitive_desc = with_bias ?
+      dnnl::matmul::primitive_desc(engine, src_desc, weights_desc, bias_desc, dst_desc, op_attr) :
+      dnnl::matmul::primitive_desc(engine, src_desc, weights_desc, dst_desc, op_attr);
+  auto primitive = dnnl::matmul(primitive_desc);
```
Do we need to check if we need to do weight reorder here?

Xia-Weiwen replied:
Thanks. Check is added.

leslie-fang-intel reviewed (changes requested):
Add few comments.

Xia-Weiwen committed:
Update on "[Quant] Add int8 linear op impl for quantization PT2E with Inductor. input is an int8 CPU tensor; weight is an int8 MdkldnnCPU tensor."


**Summary**
Add a new onednn qlinear op for quantization PT2E with Inductor. input is an int8 CPU tensor and weight is an int8 MkldnnCPU tensor.

**Test plan**
python test/test_quantization.py -k test_qlinear_pt2e

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant] Add int8 linear op impl for quantization PT2E with Inductor. input is an int8 CPU tensor; weight is an int8 MdkldnnCPU tensor."


**Summary**
Add a new onednn qlinear op for quantization PT2E with Inductor. input is an int8 CPU tensor and weight is an int8 MkldnnCPU tensor.

**Test plan**
python test/test_quantization.py -k test_qlinear_pt2e

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel commented on aten/src/ATen/native/quantized/cpu/qlinear.cpp:
```diff
@@ -900,6 +901,116 @@ at::Tensor PackedLinearWeightsOnednn:: apply_tanh(
+  tensor dst_scales_t = tensor(ideep::scale_t(1, output_scale));
+  tensor src_zp_t = tensor(ideep::zero_point_t(1, input_zero_point));
+  tensor dst_zp_t = tensor(ideep::zero_point_t(1, output_zero_point));
+  args.insert({DNNL_ARG_ATTR_SCALES | DNNL_ARG_SRC, src_scales_t});
```
Should we do the similar check as the mask here?

Xia-Weiwen replied:
Thanks. It's fixed.

Xia-Weiwen committed:
Update on "[Quant] Add int8 linear op impl for quantization PT2E with Inductor. input is an int8 CPU tensor; weight is an int8 MdkldnnCPU tensor."


**Summary**
Add a new onednn qlinear op for quantization PT2E with Inductor. input is an int8 CPU tensor and weight is an int8 MkldnnCPU tensor.

**Test plan**
python test/test_quantization.py -k test_qlinear_pt2e

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel commented:
@jerryzh168 Could you also kindly help to take a look of this PR?

Xia-Weiwen committed:
Update on "[Quant] Add int8 linear op impl for quantization PT2E with Inductor. input is an int8 CPU tensor; weight is an int8 MdkldnnCPU tensor."


**Summary**
Add a new onednn qlinear op for quantization PT2E with Inductor. input is an int8 CPU tensor and weight is an int8 MkldnnCPU tensor.

**Test plan**
python test/test_quantization.py -k test_qlinear_pt2e

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant] Add int8 linear op impl for quantization PT2E with Inductor. input is an int8 CPU tensor; weight is an int8 MdkldnnCPU tensor."


**Summary**
Add a new onednn qlinear op for quantization PT2E with Inductor. input is an int8 CPU tensor and weight is an int8 MkldnnCPU tensor.

**Test plan**
python test/test_quantization.py -k test_qlinear_pt2e

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant] Add int8 linear op impl for quantization PT2E with Inductor. input is an int8 CPU tensor; weight is an int8 MdkldnnCPU tensor."


**Summary**
Add a new onednn qlinear op for quantization PT2E with Inductor. input is an int8 CPU tensor and weight is an int8 MkldnnCPU tensor.

**Test plan**
python test/test_quantization.py -k test_qlinear_pt2e

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Xia-Weiwen commented:
@pytorchbot merge
